### PR TITLE
Enable restart/warmup workflow for GEFS simulations using state save/load

### PR DESCRIPTION
**Pull request by ZhennanShi1**
## Description

1. Updated `save_states()` and `load_states()` in dMG’s `ModelHandler` to support saving and restoring model states that include routing information.

2. Added a new documentation file:
   - `example/hydrology/README_zn.md`

3. Removed the following files:
   - `example/hydrology/example_dhbv_1_1p_gefs.ipynb`
   - `example/hydrology/HBV_1_1p.py`
   - `example/hydrology/helper.py`
 
4. Refactored the GEFS simulation workflow by splitting the original notebook into three separate notebooks:
   - `example/hydrology/example_dhbv_1_1p_cold_warm.ipynb`
   - `example/hydrology/dhbv_1_1p_zn_test.ipynb`
   - `example/hydrology/dhbv_1_1p_zn_train.ipynb`
 
5. Adapt the GEFS simulation workflow to support restart/warmup via the built-in state save/load functions.

6. Helper function is now in:  scratch/DM/generic_deltamodel/example/hydrology/tempfunctions.py

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code cleanup/refactor
- [x] Documentation update

Other (please specify):

## Checklist

- [x] Branch is up to date with master
- [ ] Updated tests or added new tests
- [ ] Tests & pre-commit hooks pass
- [ ] Updated documentation (if applicable)
- [ ] Code follows established style and conventions

### CODE REVIEW

File: example/hydrology/README_zn.md
Comment (leoglonz) [inline example/hydrology/README_zn.md:1]: should be included?

File: src/dmg/core/utils/utils.py [hunk 1]
```diff
@@ -299,7 +300,8 @@ def save_outputs(config, predictions, y_obs=None) -> None:
                 ).numpy()
 
             file_name = key + '.npy'
-            np.save(os.path.join(config['sim_dir'], file_name), out_dict)
+            # np.save(os.path.join(config['sim_dir'], file_name), out_dict)
+            np.save(os.path.join(config['output_dir'], file_name), out_dict) # zhennan changed the path directory
```
Comment (leoglonz) [inline src/dmg/core/utils/utils.py:304]: can we keep outputs in separate simulation directory? checking this

File: src/dmg/models/model_handler.py [hunk 2]
```diff
@@ -544,21 +565,33 @@ def save_states(self) -> None:
...
+            # zhennan revised the path directory
+            if 'test' in self.config['mode']: # test or test+train mode
+                save_dir = os.path.join(self.config['output_dir'], "save_states_restart")
+                os.makedirs(save_dir, exist_ok=True)
+                torch.save(state_dict, os.path.join(save_dir, "model_states.pt"))
+            else: #simulation mode 
+                print("save for simulation")
+                save_dir = os.path.join(self.config['sim_dir'], "simulation_gefs/save_states_restart")
+                os.makedirs(save_dir, exist_ok=True)
+                torch.save(state_dict, os.path.join(save_dir, "model_states.pt"))
```
Comment (leoglonz) [inline src/dmg/models/model_handler.py:590]: is It necessary to change this pathing in main code? checking for alternative

File: src/dmg/core/utils/utils.py [hunk 3]
```diff
@@ -272,7 +272,8 @@ def save_outputs(config, predictions, y_obs=None) -> None:
             c_tensor = torch.cat([d[key] for d in predictions], dim=dim)
             file_name = key + ".npy"
 
-            np.save(os.path.join(config['sim_dir'], file_name), c_tensor.numpy())
+            # zhennan changed np.save(os.path.join(config['sim_dir'], file_name), c_tensor.numpy())
+            np.save(os.path.join(config['output_dir'], file_name), c_tensor.numpy())
```
Comment (leoglonz) [inline src/dmg/core/utils/utils.py:276]: see below. checking this

**Comment by leoglonz** [comment]
@ZhennanShi1 I have integrated your changes in #94. However, I do not have the data to full evaluate the GEFS notebook `example_dhbv_1_1p_gefs.ipynb`. Let me know if you have any issues and we can address in a separate PR.

File: example/hydrology/README_zn.md
Comment (ZhennanShi1) [reply example/hydrology/README_zn.md:1]: Hi Leo, thank you for pointing it out. This should not be included.
